Governance update for Sandbox/Incubated/Graduated approach

hdamker opened the pull request:
#### What type of PR is this?

* enhancement
* project management

#### What this PR does / why we need it:

To establish the approach described within #159 and discussed on  [this wiki page](https://lf-camaraproject.atlassian.net/wiki/spaces/CAM/pages/14560178/Proposal+for+the+lifecycle+of+API+Repositories+with+Sandbox+Incubated+Graduated+Archived) there are changes to Governance documents of CAMARA necessary.

This PR changes the ProjectCharter and the ProjectStructureAndRoles:
* ProjectCharter
  * defined "Sandbox Repositories" as potentially independent of Sub Projects or Working Groups
  * clarified where necessary that Maintainers in the sense of the Charter are Maintainers of Sub Project and Working Groups (but not including the Maintainers of Sandbox Repositories)
* ProjectStructureAndRoles
  * introduced the maturity levels of Sandbox/Incubated/Graduated for API Repositories
  * described the three types of API Repositories and their technical differentiation
  * some small changes where seen necessary
  * described the responsibilities of the APIBacklog Working Group 
  * replaced "Sub Projects" with "Sub Projects and Working Groups" where appropriate

To be done in separate PRs:
* describe "Archived" repositories (after the process is described) in ProjectStructureAndRoles
* Create the document "API-Onboarding-and-Lifecycle.md" which is referenced for the processes of onboarding a new Sandbox API Repository and their Incubation and Graduation. It will replace the current API-Onboarding.md document

Out of scope of this PR:
* Further necessary changes within roles and responsibilities, especially regarding the "Participants.md" file and who are Project participants

#### Which issue(s) this PR fixes:

<!-- Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`. -->

Fixes #

#### Special notes for reviewers:



#### Additional documentation 

This section can be blank.

MarkusKuemmerle reviewed (changes requested):
The section

_A Sandbox API Repository must have at least one Codeowner and one Maintainer (could be the same person). If branch protection rules are activated (on request of the Maintainers) within the repository it must have at least two Codeowners. Branch protection rules must be get activated at latest when the first public release is created._

is wrong, as sandbox API repos don't have maintainers

Please check the section about Release Management. Do Sandbox API repos also have to comply with the work in the commonalities Working Group?
_Each Sub Project **and each Sandbox API repository** shall have a release management._ ?

Please check the sections "Maintainers" and "Changes in maintainership". Maintainers are also relevant for Working Groups. The sentence
_changing the MAINTAINERS.MD file of the respective Sub Project._
is wrong, Sub Projects don't have MAINTAINER files.

I'll create a separate issue for the clean up of "Project participants", "contributors" and the relation to PARTICIPANTS.MD.

shilpa-padgaonkar commented on ProjectStructureAndRoles.md:
```diff
@@ -2,42 +2,95 @@
-The Contributors to an API Repository are recorded by the mechanisms of GitHub in the Sub Project’s repositories. Each Contributor, Codeowner and Maintainer should also be listed in the [PARTICIPANTS.MD](./PARTICIPANTS.MD) file in the Governance repository root directory. Each API Repository should have a license file, a GOVERNANCE.MD file (pointing to the Governance repository) and subdirectorie ...
+A Sub Project should have at least three Maintainers. The Maintainers of a Sub Project are defined by the union of the Maintainers listed within the MAINTAINERS.md files of the API Repositories which are belonging to the Sub Project. It is recommended that there is an overlap between the Maintainers of the API Repositories of the Sub Project, but it is not mandated that all Maintainers of the Sub ...
+
+Each Sub Project is following for the API Repositories the Release Process as defined by the Release Management Working Group.
```
```suggestion
Each Sub Project follows the Release Process for the API Repositories as defined by the Release Management Working Group.
```

shilpa-padgaonkar commented on ProjectStructureAndRoles.md:
```diff
@@ -2,42 +2,95 @@
-A Sub Project should have at least one Maintainer. Ideally a Sub Project is managed by two or more Maintainers, depending on the size and scope of the Sub Project. The Maintainers of a Sub Project are defined by the union of the Maintainers listed within the MAINTAINERS.md files of the API Repositories which are belonging to the Sub Project. It is recommended that there is an overlap between the  ...
+The purpose of an API Repository is the definition and documentation of one or multiple (very closely related) APIs which will be always released together.
+
+An API Repository should have at least one Maintainer. Ideally an API Repository is managed by two or more Maintainers, depending on the size and scope of the API. Here, the responsibilities must be clearly agreed upon between all of the Maintainers. This includes coordinating who is responsible for which issues and pull requests. The Contributors to an API Repository are recorded by the mechanis ...
```
For repositories which are not sandbox, is it enough to have 1 Maintainer?

shilpa-padgaonkar commented on ProjectStructureAndRoles.md:
```diff
@@ -2,42 +2,95 @@
-**Working Groups**
+A Sandbox API Repository can belong to a CAMARA Sub Project or exist initially independent. If the Sandbox API Repository belongs to a Sub Project it links within the README.md to the Sub Project wiki page. Mailing list, meetings will be for these Sandbox API Repositories organized in context of the Sub Project. A Sandbox API Repository which is not part of an existing Sub Project will get a Wiki ...
+
+A Sandbox API Repository must have at least one Codeowner and one Maintainer (could be the same person). If branch protection rules are activated (on request of the Maintainers) within the repository it must have at least two Codeowners. Branch protection rules must be get activated at latest when the first public release is created.
```
```suggestion
A Sandbox API Repository must have at least one Codeowner and one Maintainer (could be the same person). If branch protection rules are activated (on request of the Maintainers) within the repository it must have at least two Codeowners. Branch protection rules must get activated at latest when the first public release is created.
```

shilpa-padgaonkar commented on ProjectStructureAndRoles.md:
```diff
@@ -2,42 +2,95 @@
+
+**Incubated API Repositories**
+
+An Incubated API Repository is meant for the further evolution of an API which has been implemented already by at least one operator, fits into the CAMARA API portfolio and gained sufficient support within the CAMARA community. For details of the Incubation process see [API-Onboarding-and-Lifecycle.md](./API-Onboarding-and-Lifecycle.md).
```
```suggestion
An Incubated API Repository is meant for the further evolution of an API which has been implemented already by at least one operator, fits into the CAMARA API portfolio and has gained sufficient support within the CAMARA community. For details of the Incubation process see [API-Onboarding-and-Lifecycle.md](./API-Onboarding-and-Lifecycle.md).
```

shilpa-padgaonkar commented on ProjectStructureAndRoles.md:
```diff
@@ -65,9 +118,11 @@ More details can be found in the [CONTRIBUTING.MD](https://github.com/camaraproj
+The decision about new Sandbox API Repositories will be taken by the Working Group. Decision about Incubation, Graduation, and onboarding of new Sub Projects will be taken by the TSC. Changes of existing Sub Projects have to be approved by the involved Sub Projects.
 
-The API onboarding and management in CAMARA is described in [API-onboarding.md](./documentation/API-onboarding.md).
+The API onboarding and API lifecycle process in CAMARA will described in [API-Onboarding-and-Lifecycle.md](./API-Onboarding-and-Lifecycle.md).
```
```suggestion
The API onboarding and API lifecycle process in CAMARA are described in [API-Onboarding-and-Lifecycle.md](./API-Onboarding-and-Lifecycle.md).
```

shilpa-padgaonkar commented on ProjectStructureAndRoles.md:
```diff
@@ -48,7 +101,7 @@ In the Project the “Fork and pull model” is used. Changes and contributions
 2 Either the issue creator or one of the Sub Project Maintainers then shall assign the issue to one of the Contributors of the Sub Project.
 
-3 This Contributor shall create a fork (if necessary), shall create a new branch, and shall copy the branch link into the issue. For the naming of the branch a clear and concrete description shall be used.
+3 This Contributor shall create a fork (if not yet done before), shall create a new branch within the fork, and shall copy the branch link into the issue. For the naming of the branch a clear and concrete description shall be used.
```
Would copying the link in the issue be necessary? If we use the PR template and the #fixes tag, then this copying step is not needed. Probably not relevant for this PR but added this as this statement was part of the change.

hdamker replied:
You are right, by using the "fixes" logic of GitHub, the PR will get automatically copied into the issue. But it is also possible to Lin the branch already earlier (there is a field within a GitHub issue for that, which is e.g. filled if the branch will be created out of the issue). Making this optional could be the scope of a separate PR.

Kevsy commented on ProjectCharter.md:
```diff
@@ -4,15 +4,16 @@
-- **Maintainer**: is leader of an individual Sub Project or Working Group who has been named in the MAINTAINERS.MD files of the respective Sub Project.
-- **Codeowner**: a committer of an individual Sub Project who has been named in the CODEOWNERS files of the Sub Project. Codeowners merge code and accept pull requests.
+- **Sandbox Repositories** can exist independent of a Sub Project or Working Group, e.g. during the initial development phase of an API. 
+- **Maintainer**: is leader of an individual Sub Project or Working Group who has been named in the MAINTAINERS.MD files of one of the repositories of the Sub Project or Working Group. 
```
```suggestion
- **Maintainer**: is a leader of an individual Sub Project or Working Group who has been named in the MAINTAINERS.MD files of one of the repositories of the Sub Project or Working Group. 
```

Kevsy commented on ProjectCharter.md:
```diff
@@ -141,12 +142,12 @@ is created.
 -	1 non-voting representative of the TMForum
 
-Any Maintainer may nominate themselves or another Maintainer of any Sub Project for consideration as a Sub Project representative. The Maintainers will vote using ranked choice voting (or another method approved by the TSC) and those 10 nominees receiving the highest number of votes will be elected as Sub Project representatives.
+Any Maintainer may nominate themselves or another Maintainer of any Sub Project or Working Group for consideration as a Sub Project representative. The Maintainers will vote using ranked choice voting (or another method approved by the TSC) and those 10 nominees receiving the highest number of votes will be elected as Sub Project representatives.
```
```suggestion
Any Maintainer may nominate themselves or another Maintainer of any Sub Project or Working Group for consideration as a Sub Project representative. The Maintainers will vote using ranked choice voting (or another method approved by the TSC) and those 10 nominees receiving the highest number of votes will be elected to the TSC as Sub Project representatives.
```

hdamker commented:
@MarkusKuemmerle Thanks for your review!

> The section
> 
> _A Sandbox API Repository must have at least one Codeowner and one Maintainer (could be the same person). If branch protection rules are activated (on request of the Maintainers) within the repository it must have at least two Codeowners. Branch protection rules must be get activated at latest when the first public release is created._
> 
> is wrong, as sandbox API repos don't have maintainers

I did changes to make maintainers for Sandbox API Repositories optional within https://github.com/camaraproject/Governance/pull/161/commits/680d7cf02369a456efdbe9af153ac734b17d7dac 

But we have to differentiate here between two cases:
* A Sandbox API Repository which belongs to an existing Sub Project. Here the Sub Project has a defined set of Maintainers who will oversee also the Sandbox Repository. The existing Maintainers may nominate or approve already Maintainers for the Sandbox Repository which are then noted within the MAINTAINERS. 
* An independent Sandbox API Repository which is not (yet) part of a Sub Project. Here you are right, they can't have Maintainers in the sense of the ProjectCharter ("leaders of a Sub Project") as there is no Sub Project yet. But the Sandbox phase has also the purpose to attract contributors who are willing to take the role of Maintainers, e.g. after Incubation. Therefore I formulated the role of MAINTAINERS.md file within an independent Sandbox API Repository in this sense: as a list of participants committed to take the role after Incubation or adoption by a Sub Project.

> 
> Please check the section about Release Management. Do Sandbox API repos also have to comply with the work in the commonalities Working Group? _Each Sub Project **and each Sandbox API repository** shall have a release management._ ?
> 
I tried to clarify this in https://github.com/camaraproject/Governance/pull/161/commits/197a6a6a974ef69a7355cbb443ebaeb83f392587 and https://github.com/camaraproject/Governance/pull/161/commits/680d7cf02369a456efdbe9af153ac734b17d7dac mainly by the sentence within the definition of the Release Management working group: "_All API and Working Group Repositories must comply with the work in the Release Management Working Group for releases and pre-releases._"

> Please check the sections "Maintainers" and "Changes in maintainership". Maintainers are also relevant for Working Groups. The sentence _changing the MAINTAINERS.MD file of the respective Sub Project._ is wrong, Sub Projects don't have MAINTAINER files.
> 
Addressed within https://github.com/camaraproject/Governance/pull/161/commits/680d7cf02369a456efdbe9af153ac734b17d7dac and https://github.com/camaraproject/Governance/pull/161/commits/c58a241edcf6b06ed9a8e2e522c11cd362787a57 by replacing "Sub Projects" by "Sub Projects and Working Groups" wherever appropriate, correcting the sentence and adding the clarification that changes of maintainership of Working Groups are the responsibility of the TSC. I might still have missed some instances, please check and suggest needed changes.

> I'll create a separate issue for the clean up of "Project participants", "contributors" and the relation to PARTICIPANTS.MD.

Thanks!

hdamker commented:
Opening the PR for review, as the initial comments are (hopefully) addressed.
The "API-Onboarding-and-Lifecycle.md" document will be separate PR later.

jgarciahospital commented on ProjectStructureAndRoles.md:
```diff
@@ -2,42 +2,95 @@
 **Sub Projects**
 
-The CAMARA Project is organized primarily into Sub Projects. Each Sub Project is composed of Project participants from multiple companies and organizations, with a common purpose of advancing the Sub Project with respect to a specific Service API topic, for example ‘quality on demand’ or ‘localization’. Our goal is to enable distributed decision making and code ownership. This will be done by pro ...
+Sub Projects are the primary structure of the CAMARA Project. Each Sub Project is composed of Project participants from multiple companies and organizations, with a common purpose of advancing the Sub Project with respect to a specific Service API topic, for example ‘quality on demand’ or ‘localization’. Our goal is to enable distributed decision making and code ownership. This will be done by pr ...
```
Should we include here the term family? As it has evolved from its creation, right now families are mostly parallelizing the subgroup concept, a group of APIs with a relation among them but separated in different repositories to ensure that there is not cross-dependancy in the specification. Location, qod... are good examples for API families.

bigludo7 replied:
I'm a bit skeptical to re introduce the family term. We had discussion on the past but is was difficult to define and we get ride of it. Sub project term is just go enough for me.

hdamker replied:
I'm here with Ludovic, we shouldn't use the term as an official term going forward as we have the structure elements of API Repository and Sub Projects with the "common purpose of advancing the Sub Project with respect to a specific Service API topic". We will get in addition the API Portfolio categories to structure the overall set of APIs within the communication towards API Consumers.

jgarciahospital commented on ProjectStructureAndRoles.md:
```diff
@@ -2,42 +2,95 @@
+A Sandbox API Repository can be created independent of a CAMARA Sub Project or in context of an existing CAMARA Sub Project. It is the first stage for every new API proposal developed in CAMARA, for details see [API-Onboarding-and-Lifecycle.md](./API-Onboarding-and-Lifecycle.md).
 
-Working Groups are documented in a separate section of [wiki.camaraproject.org](https://wiki.camaraproject.org) and can have one or multiple GitHub repositories following the API Repository template (the rules for API Repositories apply). In the Readme.md file the repositories of Working Groups are marked with a "Working Group" badge.
+A Sandbox API Repository can belong to a CAMARA Sub Project or exist initially independent as an independent API Repository. 
```
Duplicated information with line 45

hdamker replied:
Addressed in https://github.com/camaraproject/Governance/pull/161/commits/09911571145eb88a3dd3896a0151ac9f19f6b4c5

jgarciahospital commented on ProjectStructureAndRoles.md:
```diff
@@ -2,42 +2,95 @@
-As mentioned before the contributions to the Sub Projects were documented in the respective GitHub repositories of the Sub Projects, and the contributions to Working Groups were documented in the GitHub repositories of the Working Groups. All contributions beyond that, esp. contributions to the Project itself or the TSC are documented in the GitHub repository “Governance”. Additional repositories ...
+* An independent Sandbox API Repository which is not part of an existing Sub Project will get a page in a separate section of the [CAMARA wiki](https://lf-camaraproject.atlassian.net/), a preliminary mailing list, a Slack channel, and (on request) a Zoom meeting to be able to communicate with the community and attract contributors. The MAINTAINERS.md file of an Independent Sub Project may already ...
+
+A Sandbox API Repository must have at least one Codeowner. If branch protection rules are activated within the repository it must have at least two Codeowners. Branch protection rules must get activated at latest when the first public release is created.
```
I'd delete the branch protection description here, since it's agreed that this not applies to sandbox repos.

hdamker replied:
My idea here was that a public release should be protected, but I agree that this would restrict the work within a Sandbox. My view is that latest a stable API version should be maintained with branch protection, but that we can also define within Release Management. 

I changed the paragraph in https://github.com/camaraproject/Governance/pull/161/commits/d302e9501ae99f4d0d8ba729ef1e128282338c86 to:

> A Sandbox API Repository must have at least one Codeowner. If the Sandbox API Repository has two or more Codeowners the activation of branch protection rules is recommended.

jgarciahospital commented on ProjectStructureAndRoles.md:
```diff
@@ -2,42 +2,95 @@
-For ease of use there are templates for the two types of Sub Project repositories (API and Provider Implementation Repositories), containing all necessary files which can be cloned and adapted. When cloning it has to be ensured that also the branch protection rules are active in the new repositories. For Working Group repositories the API Repository template should be used and adapted.
+An Incubated API Repository has the same structure as a Sandbox API Repository, with the following additional requirements: 
+* An Incubated API Repository must belong to a CAMARA Sub Project (either an existing one or a new Sub Project created as part of the Incubation process). 
+* The repository should have at least three Maintainers (out of the group of Maintainers of the Sub Project it belongs to). 
```
is it necessary to specify if the maintainers need to belong to different companies here? we discussed about this in the draft, both for maintainers support and market adoption rules.

bigludo7 replied:
Agree with @jgarciahospital that this point should be added.

hdamker replied:
Agree, addressed in f7082010fabe23a6a64817685672acdba3f5a05f

bigludo7 reviewed (commented):
Look globally good to me. Share small comment that Jorge about indication that maintainers for incubated projects should be from at least 2 distinct companies.

jgarciahospital commented on ProjectStructureAndRoles.md:
```diff
@@ -2,42 +2,95 @@
 
-**Working Groups**
+A Sandbox API Repository is meant for the rapid development of a new API proposal through one or multiple initial API versions (0.x.y) until the first public release of an initial or stable API Version within the context of a CAMARA Meta-release.
 
```
Even when sandbox proposals are not detaily analized that will not occur until they have progressed and have a clearer scope and technical definition, We need to include basic rules that ensures that proposals included as sandbox, are still following the spirit of CAMARA.
```suggestion
Sandbox API Repositories 
```

hdamker replied:
I agree with the intention of the suggested change, but would avoid to rephrase the scope (defined in the ProjectCharter) or technical API design guidelines in this document. Instead I added with 3b49da2449854875a7f8419f9fc3b58af6f1d1ce the following sentence:

> The work within Sandbox API Repositories - like in all API repositories - shall follow the governance and guidelines of CAMARA, including but not limited to the ProjectCharter, the Code of Conduct, the Contribution guidelines, the API Design Guidelines of Commonalities and the Release Process defined in Release Management.

jgarciahospital replied:
Agreed, ok.

hdamker commented:
@shilpa-padgaonkar @MarkusKuemmerle @jgarciahospital @bigludo7 I tried to address all review comments with appropriate changes. It would be great if you can review these changes, resolve the conversation you have opened if sufficiently answered and approve finally the PR.

bigludo7 reviewed (dismissed):
LGTM
Thanks a lot Herbert

MarkusKuemmerle commented:
I've reviewed the current version, please find my proposals in the DOC file attached, amber marked texts.
Mainly typos and clarifications. The only open point is if we want to inform TSC about every member change of a working group. That will not work in reality...
[Project Structure and Roles.docx](https://github.com/user-attachments/files/17880446/Project.Structure.and.Roles.docx)

hdamker commented:
> I've reviewed the current version, please find my proposals in the DOC file attached, amber marked texts.
> Mainly typos and clarifications.
> [Project Structure and Roles.docx](https://github.com/user-attachments/files/17880446/Project.Structure.and.Roles.docx)

Thanks @MarkusKuemmerle for the detailed review. 
* Typos and minor edits addressed in 9b122a5b4326ce84ea5fc8426aa323441364ba8c
* Clarification that rules for API Repositories shall apply for sandboxed, incubated, and graduated API repositories added in 1d46ff3b982f77850b21212b729140049d599cae
* Replaced "Sub Projects and Working Groups" with "Repositories" where appropriate and added clarification that Independent Sandbox Repositories don't have Maintainers in the sense of the Project Governance in 00431c088776895fb684134db81797ccdde19f63

> The only open point is if we want to inform TSC about every member change of a working group. That will not work in reality...
The procedure is not about change of every member but only about changes within the team of Maintainers, and the TSC is supervising the Working Groups. We can reconsider all these procedures in a separate issue if needed.

hdamker commented:
@MarkusKuemmerle please review if your requested changes are addressed.
@bigludo7 please review the changes done to address the review comments in https://github.com/camaraproject/Governance/pull/161#issuecomment-2504396919
@camaraproject/tsc-participants please review and approve as agreed in TSC on Nov 21st until Nov 28th.

bigludo7 reviewed (approved):
LGTM
Thanks Herbert

hdamker commented:
Thanks for the reviews! As agreed within TSC on Nov 21st I will merge the PR now. We might encounter smaller needs for change while we operationalize the concept (e.g. Wiki structure), but these can be done in a new PR.
